proposed/0008 MPID Messaging.md

```diff
@@ -0,0 +1,295 @@
+    1. support Delete (for StructuredData only)
+    1. address relocation (allows Client to use a fixed MPID address to connect to the network and allows "push" model - not required for "pull" model)
+
+1. Client
+    1. Put `MpidMessage`
+    1. Get all `MpidHeader`s (pull model) or accept all/single `MpidHeader` (push model)
+    1. Get `MpidMessage`.  This shall also include the work of removing corresponding `MpidHeader`s
+    1. Delete `MpidMessage`
+    1. address relocation (allows Client to use a fixed MPID address to connect to the network and allows "push" model - not required for "pull" model)
+    1. address registration (not required if address relocation is used)
+
+# Drawbacks
+
+None identified, other than increased complexity of Vault and Client codebase.
+
+# Alternatives
+
+No other in-house alternatives have been documented as yet.
+
+If we were to _not_ implement some form of secure messaging system, a third party would be likely to implement a similar system using the existing interface to the SAFE Network.  This would be unlikely to be as efficient as the system proposed here, since this will be "baked into" the Vault protocols.
+
+We also have identified a need for some form of secure messaging in order to implement safecoin, so failure to implement this RFC would impact on that too.
+
+# Unresolved Questions
+
+1. Should we use the "push" or "pull" model?  At the moment, Qi and Fraser prefer the pull model on the basis that it involves less work: "address registration" or "address relocation" procedures are not required as the MPID Manager is always responding to requests from the MPID Client directly.  Furthermore, changing from the pull to push model at a later stage would appear to involve less wasted coding effort than vice versa.
+
+1. `MpidMessage` and `MpidHeader` are wrapped in `StructuredData` instances in order to allow Delete requests on them and to allow them to be given different tags (otherwise they could be ImmutableData, since they don't mutate).  Should Routing, which already has to define these two types, be made "aware" of these two types?  (i.e. should they get added to the [`::routing::data::Data` enum][6])
+
+    Identified pros:
+    - less wrapping/parsing, so simpler for Vaults and Clients to deal with
+    - more efficient (smaller messages to transmit and store)
+    - no need for Routing to publicly expose `MPID_MESSAGE_TAG` or `MPID_HEADER_TAG`
+
+    Identified cons:
+    - increased complexity in Routing
+    - would need to add the sender's public key to the header Put flow
+
+    Qi prefers using `StructuredData`, Fraser prefers not using `StructuredData` unless the effort required by Routing to accommodate the new types is significant.
+
+1. By having the `recipient` field inside `MpidMessage::signed_header`, a `sign::verify` call is required at every hop to parse the header and identify the recipient.  An alternative is to duplicate the `recipient` field in the `MpidMessage` itself.
```

Comment:
An alternative is to put the `recipient` field in the `MpidMessage` instead.  (not a duplication but a move)
this will eliminates the concern of increased message size and need to carry out consistency checks.
The only identified con is: `recipient` is not protected by signed_header
However this will be easily overcome by have signed_body = sign(recipient+content)
I prefer to move the `recipient` from `MpidHeader` to `MpidMessage`.